### PR TITLE
POSIX compliant batchrename plugin

### PR DESCRIPTION
**Pull request by KlzXS**
Note that `mktemp` is not POSIX.

Please test this, cover some edge cases. Try using filenames with `'`. I wouldn't recommend using it with important files until tested as it does `mv` files around.

### CODE REVIEW

**Comment by KlzXS** [comment]
CI timed out.

**Comment by jarun** [comment]
Restarted CI.

Can you modify it so it asks whether to use selection (if something is selected) or open entries in current dir?

File: plugins/batchrename [hunk 1]
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env sh
+
+# Description: An almost fully POSIX compliant way to batch rename files
+#
+# Shell: POSIX compliant
+# Author: KlzXS
+
+EDITOR="${EDITOR:-vim}"
```
Comment (jarun) [inline plugins/batchrename]: Please use `vi`.

File: plugins/batchrename [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+TMPDIR="${TMPDIR:-/tmp}"
+
+selection=${XDG_CONFIG_HOME:-$HOME/.config}/nnn/.selection
```
Comment (jarun) [inline plugins/batchrename]: If there's a selection (size of .selection file is on-0) prompt user for selection or current dir.
In case there's no selection (size of .selection file is 0), open the entries in current dir directly.

File: plugins/batchrename [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+exit_status=0
+
+dst_file=$(mktemp "$TMPDIR/.nnnXXXXXX")
```
Comment (jarun) [inline plugins/batchrename]: Probably have a switch to include hidden files as well. By default we can go non-hidden only.

File: plugins/batchrename [hunk 4]
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env sh
+
+# Description: An almost fully POSIX compliant way to batch rename files
```
Comment (jarun) [inline plugins/batchrename]: An almost fully POSIX compliant batch file renamer

**Comment by KlzXS** [comment]
> Can you modify it so it asks whether to use selection (if something is selected) or open entries in current dir?

I though you wanted to setup selection in `batch_rename()` and then call the plugin? I can modify it.

File: plugins/batchrename (hunk 3 above)
Comment (KlzXS) [reply plugins/batchrename]: This is in regards to editing the current directory entries? What should the switch be, I don't really like `-h`, that looks too much like "help".

Comment (jarun) [reply plugins/batchrename]: Not that.

Have a variable at the top like:

    INCLUDE_HIDDEN=0

If user makes it 1, include the hidden files **in the dir** (doesn't apply to selection anyway) as well.

**Comment by jarun** [comment]
> I though you wanted to setup selection in batch_rename() and then call the plugin? I can modify it.

We'll have to make it behave like other plugins. But the key to run this would remain `r`. So select in `nnn` and press `r`. If there's a selection, the script prompts...

**Comment by jarun** [comment]
This is an excellent script BTW. I don't know of any POSIX-compliant _intelligent_ shell script to batch rename!

**Comment by jarun** [comment]
Thank you so much!

@0xACE @maximbaz let's test this!

**Comment by maximbaz** [comment]
Thanks for the ping!

Out of curiosity, why did you decide to move this functionality out of the core and into a plugin?

> We'll have to make it behave like other plugins. But the key to run this would remain r. So select in nnn and press r. If there's a selection, the script prompts...

Just as a heads-up, until now plugins were optional, and for example they are not being shipped as part of distro packages, so unless we address this, thousands of users will require manual intervention after nnn upgrade in order to fix the rename functionality.

**Comment by jarun** [comment]
The core batch rename functionality is limited.

We can still have the core functionality as it is and invoke the plugin if it's available.

**Comment by maximbaz** [comment]
If you want, we can also just start packaging plugins, so that you dont have to maintain two implementations of the same functionality - it should be as easy as adding one line in `install` target in Makefile 👍

**Comment by jarun** [comment]
... that and the current renamer plugin with external deps goes.

Plugins are installed per-user, only if they want. Not planning to install them system-wide.
